File: Sources/GUIForCLIWindows.Core/ActionModels.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace GUIForCLIWindows.Core;

public sealed record ConfigSettingSpec
{
    [JsonPropertyName("id")]
    public string Id { get; init; } = "";

    [JsonPropertyName("kind")]
    public string Kind { get; init; } = "";

    [JsonPropertyName("key")]
    public string Key { get; init; } = "";

    [JsonPropertyName("value")]
    public string? Value { get; init; }

    [JsonPropertyName("label")]
    public string Label { get; init; } = "";

    [JsonPropertyName("placeholder")]
    public string? Placeholder { get; init; }

    [JsonPropertyName("tooltip")]
    public string? Tooltip { get; init; }

    [JsonPropertyName("options")]
    public List<ControlOption> Options { get; init; } = [];
}

public sealed record ConfigFileSpec
{
    [JsonPropertyName("path")]
    public string Path { get; init; } = "";
}

public sealed record ActionSpec
{
    [JsonPropertyName("id")]
    public string Id { get; init; } = "";

    [JsonPropertyName("title")]
    public string Title { get; init; } = "";

    [JsonPropertyName("tooltip")]
    public string? Tooltip { get; init; }

    [JsonPropertyName("visibleWhen")]
    public List<ActionConditionSpec> VisibleWhen { get; init; } = [];

    [JsonPropertyName("disabledWhen")]
    public List<ActionConditionSpec> DisabledWhen { get; init; } = [];

    [JsonPropertyName("disabledTooltip")]
    public string? DisabledTooltip { get; init; }

    [JsonPropertyName("destructive")]
    public bool Destructive { get; init; }

    [JsonPropertyName("precheck")]
    public ActionPrecheckSpec? Precheck { get; init; }

    [JsonPropertyName("command")]
    public CommandSpec Command { get; init; } = new();

    [JsonPropertyName("confirm")]
    public ConfirmationSpec? Confirm { get; init; }
}

public sealed record ConfirmationSpec
{
    [JsonPropertyName("title")]
    public string Title { get; init; } = "";

    [JsonPropertyName("message")]
    public string Message { get; init; } = "";

    [JsonPropertyName("confirmButtonTitle")]
    public string ConfirmButtonTitle { get; init; } = "";

    [JsonPropertyName("cancelButtonTitle")]
    public string CancelButtonTitle { get; init; } = "";

    [JsonPropertyName("requiredText")]
    public string? RequiredText { get; init; }

    [JsonPropertyName("prompt")]
    public string? Prompt { get; init; }
}

public sealed record ActionConditionSpec
{
    [JsonPropertyName("placeholder")]
    public string Placeholder { get; init; } = "";

    [JsonPropertyName("exists")]
    public bool? Exists { get; init; }

    [JsonPropertyName("equals")]
    public string? EqualTo { get; init; }

    [JsonPropertyName("notEquals")]
    public string? NotEquals { get; init; }

    [JsonPropertyName("in")]
    public List<string> In { get; init; } = [];

    [JsonPropertyName("notIn")]
    public List<string> NotIn { get; init; } = [];

    [JsonPropertyName("lessThan")]
    public string? LessThan { get; init; }

    [JsonPropertyName("lessThanOrEqual")]
    public string? LessThanOrEqual { get; init; }

    [JsonPropertyName("greaterThan")]
    public string? GreaterThan { get; init; }

    [JsonPropertyName("greaterThanOrEqual")]
    public string? GreaterThanOrEqual { get; init; }
}

public sealed record CommandSpec
{
    [JsonPropertyName("executable")]
    public string Executable { get; init; } = "";

    [JsonPropertyName("arguments")]
    public List<string> Arguments { get; init; } = [];

    [JsonPropertyName("optionalArguments")]
    public List<List<string>> OptionalArguments { get; init; } = [];
}

public sealed record DataSourceSpec
{
    [JsonPropertyName("path")]
    public string Path { get; init; } = "";

    [JsonPropertyName("arguments")]
    public List<string> Arguments { get; init; } = [];

    [JsonPropertyName("workingDirectory")]
    public string? WorkingDirectory { get; init; }

    [JsonPropertyName("environment")]
    public Dictionary<string, string> Environment { get; init; } = [];
}

public sealed record ActionPrecheckSpec
{
    [JsonPropertyName("diskSpaceGB")]
    public string? DiskSpaceGB { get; init; }

    [JsonPropertyName("diskSpacePath")]
    public string? DiskSpacePath { get; init; }

    [JsonPropertyName("warningMessage")]
    public string? WarningMessage { get; init; }
}

File: exp-platform/dart/flutter/lib/src/models/action_models.dart
part of '../models.dart';

class ActionSpec {
  ActionSpec({
    required this.id,
    required this.title,
    required this.command,
    this.tooltip,
    this.iconName,
    this.iconEmoji,
    this.iconOnly = false,
    this.role = 'primary',
    this.destructive = false,
    this.visibleWhen = const [],
    this.disabledWhen = const [],
    this.disabledTooltip,
    this.precheck,
    this.confirm,
  });

  final String id;
  final String title;
  final String? tooltip;
  final String? iconName;
  final String? iconEmoji;
  final bool iconOnly;
  final String role;
  final bool destructive;
  final List<ActionConditionSpec> visibleWhen;
  final List<ActionConditionSpec> disabledWhen;
  final String? disabledTooltip;
  final ActionPrecheckSpec? precheck;
  final ActionConfirmationSpec? confirm;
  final CommandSpec command;

  factory ActionSpec.fromJson(Map<String, Object?> json) => ActionSpec(
        id: stringValue(json['id']),
        title: stringValue(json['title']),
        tooltip: optionalString(json['tooltip']),
        iconName: optionalString(json['iconName']) ??
            optionalString(json['systemImage']),
        iconEmoji: optionalString(json['iconEmoji']),
        iconOnly: json['iconOnly'] == true,
        role: optionalString(json['role']) ?? 'primary',
        destructive:
            json['destructive'] == true || json['role'] == 'destructive',
        visibleWhen: listOfMaps(json['visibleWhen'])
            .map(ActionConditionSpec.fromJson)
            .toList(),
        disabledWhen: listOfMaps(json['disabledWhen'])
            .map(ActionConditionSpec.fromJson)
            .toList(),
        disabledTooltip: optionalString(json['disabledTooltip']),
        precheck: json['precheck'] is Map<String, Object?>
            ? ActionPrecheckSpec.fromJson(
                json['precheck']! as Map<String, Object?>)
            : null,
        confirm: json['confirm'] is Map<String, Object?>
            ? ActionConfirmationSpec.fromJson(
                json['confirm']! as Map<String, Object?>)
            : null,
        command: json['command'] is Map<String, Object?>
            ? CommandSpec.fromJson(json['command']! as Map<String, Object?>)
            : const CommandSpec(executable: '', arguments: []),
      );
}

class ActionConditionSpec {
  const ActionConditionSpec({
    required this.placeholder,
    this.equals,
    this.notEquals,
    this.inValues = const [],
    this.notInValues = const [],
    this.exists,
    this.lessThan,
    this.lessThanOrEqual,
    this.greaterThan,
    this.greaterThanOrEqual,
  });

  final String placeholder;
  final String? equals;
  final String? notEquals;
  final List<String> inValues;
  final List<String> notInValues;
  final bool? exists;
  final String? lessThan;
  final String? lessThanOrEqual;
  final String? greaterThan;
  final String? greaterThanOrEqual;

  factory ActionConditionSpec.fromJson(Map<String, Object?> json) =>
      ActionConditionSpec(
        placeholder: stringValue(json['placeholder']),
        equals: optionalString(json['equals']),
        notEquals: optionalString(json['notEquals']),
        inValues: listOfStrings(json['in']),
        notInValues: listOfStrings(json['notIn']),
        exists: json['exists'] is bool ? json['exists']! as bool : null,
        lessThan: optionalString(json['lessThan']),
        lessThanOrEqual: optionalString(json['lessThanOrEqual']),
        greaterThan: optionalString(json['greaterThan']),
        greaterThanOrEqual: optionalString(json['greaterThanOrEqual']),
      );
}

class ActionPrecheckSpec {
  const ActionPrecheckSpec({
    this.diskSpaceGB,
    this.diskSpacePath,
    this.warningMessage,
  });

  final String? diskSpaceGB;
  final String? diskSpacePath;
  final String? warningMessage;

  factory ActionPrecheckSpec.fromJson(Map<String, Object?> json) =>
      ActionPrecheckSpec(
        diskSpaceGB: optionalString(json['diskSpaceGB']),
        diskSpacePath: optionalString(json['diskSpacePath']),
        warningMessage: optionalString(json['warningMessage']),
      );
}

class ActionConfirmationSpec {
  const ActionConfirmationSpec({
    required this.title,
    this.message,
    this.confirmButtonTitle = 'Continue',
    this.cancelButtonTitle = 'Cancel',
    this.requiredText,
    this.prompt,
  });

  final String title;
  final String? message;
  final String confirmButtonTitle;
  final String cancelButtonTitle;
  final String? requiredText;
  final String? prompt;

  factory ActionConfirmationSpec.fromJson(Map<String, Object?> json) =>
      ActionConfirmationSpec(
        title: stringValue(json['title']),
        message: optionalString(json['message']),
        confirmButtonTitle:
            optionalString(json['confirmButtonTitle']) ?? 'Continue',
        cancelButtonTitle:
            optionalString(json['cancelButtonTitle']) ?? 'Cancel',
        requiredText: optionalString(json['requiredText']),
        prompt: optionalString(json['prompt']),
      );
}
